src/hotspot/share/gc/shenandoah/shenandoahConcurrentGC.cpp

```diff
@@ -718,8 +861,14 @@ void ShenandoahEvacUpdateCleanupOopStorageRootsClosure::do_oop(oop* p) {
   const oop obj = RawAccess<>::oop_load(p);
   if (!CompressedOops::is_null(obj)) {
     if (!_mark_context->is_marked(obj)) {
-      shenandoah_assert_correct(p, obj);
-      ShenandoahHeap::atomic_clear_oop(p, obj);
+      shenandoah_assert_generations_reconciled();
+      if (_heap->is_in_active_generation(obj)) {
+        // Here we are asserting that an unmarked from-space object is 'correct'. There seems to be a legitimate
+        // use-case for accessing from-space objects during concurrent class unloading. In all modes of Shenandoah,
+        // concurrent class unloading only happens during a global collection.
+        shenandoah_assert_correct(p, obj);
```

Comment:
I noticed this before and removed that assert here: https://github.com/openjdk/jdk/commit/b9d49dcef22ab81a087d890bbac0329a5244a2ef#diff-3876d66a8561c2fe5826c6217603b76efde78847ec46b11936af52158f58c27cR721